docs/proposals/project-scoped-repository-enhancements.md

```diff
@@ -0,0 +1,124 @@
+perspective is to add an `appProject` parameter when retrieving or deleting a repository credential. To preserve backwards 
+compatibility this option is optional and would only be a required parameter if multiple repository credentials are 
+found for the same URL.
+
+Finally, we need to change the way the cache keys for the repository paths are generated in the repo-server 
+(see `Security Considerations`). 
+
+### Security Considerations
+
+* Special care needs to be taken in order not to inadvertently expose repository credentials belonging to other `AppProject`s.
+Access to repositories are covered by RBAC checks on the project, so we should be good.
+* We need to change how the cache keys for the checked out repository paths are generated on the repo-server side, the 
+reason being that we do not want separate `AppProject`s sharing the same paths of sources which have been downloaded. 
+With this change there is a potential for multiple `AppProject`s to have rendered/downloaded different manifests due to 
+having different sets of credentials, so to mitigate that we need to check out a separate copy of the repository per 
+`AppProject`.
+
+### Risks and Mitigations
+
+### Upgrade / Downgrade Strategy
+
+When upgrading no changes need to happen - the repository credentials will work as before. On the other hand, when 
+downgrading to an older version we need to consider that the existing order in which multiple credentials with the same
+URL gets returned is undefined. This means that deleting the credentials before downgrading to an older version would be
+advisable.
+
+## Drawbacks
+
+* It will be more difficult to reason about how a specific repository credential gets selected. There could be scenarios 
+where a repository has both a global repository credential and a scoped credential for the project to which the 
+application belongs.
+* There will be more secrets proliferating in the `argocd` namespace. This has the potential to increase maintenance burden
+to keeping said secrets safe, and it also makes it harder to have a bird's eye view from an Argo CD admin's perspective.
+* Depending on the number of projects making use of distinct credentials for the same repository URL, loading the correct 
+credentials from the repository secrets has the potential to scale linearly with the number of app projects (in the worst case 
+scenario we would need to loop through all the credentials before finding the correct credential to load). This is likely 
+a non-issue in practice.
+* Also depending on the number of projects making use of distinct credentials for the same repository URL, this will 
+imply that for each `AppProject` sharing the same repository URL, a separate copy of the repository will be checked out.
+This has potential implications in terms of memory consumption, sync times, CPU load times etc. This is something 
+which an Argo CD admin will need to be mindful of.
```

Comment:
```suggestion
about which an Argo CD admin will need to be mindful.
```

Reply:
I'm guessing that should be

```suggestion
which an Argo CD admin will need to be mindful about.
```

?

Reply:
I'm trying to avoid ending the sentence with a preposition.

Reply:
> This is something about which an Argo CD admin will need to be mindful.

...sounds wrong to me, as if Yoda is the author of that sentence. How about 
```suggestion
of which an Argo CD admin will need to be mindful.
```

?